ciao-controller/api/api.go

```diff
@@ -96,9 +98,18 @@ func errorResponse(err error) Response {
 type Handler struct {
 	*Context
 	Handler func(*Context, http.ResponseWriter, *http.Request) (Response, error)
+	Priv    bool
 }
 
 func (h Handler) ServeHTTP(w http.ResponseWriter, r *http.Request) {
+	// check whether we should send permission denied for this route.
+	if h.Priv {
+		_, ok := context.GetOk(r, service.PrivKey)
+		if !ok {
+			http.Error(w, http.StatusText(http.StatusUnauthorized), http.StatusUnauthorized)
```

Comment:
I'm a bit confused. Does the code below still not get called? And thus although you've sent a 403 and side effects will still happen by the call into h.Handler(h.Context, w, r) ?